### DISCUSSION TITLE
Out-of-Sample Predictions

### DISCUSSION THREAD

**Discussion by louismagowan** [pymc-labs/pymc-marketing#392] [Q&A] [answered]
Does anyone know how to get out-of-sample or test-set predictions with a fitted DelayedSaturatedMMM model?

When I try to predict posterior samples using test data I receive dimension errors (since X_test has fewer rows than X_train)

e.g. this error `ValueError: The 'channel_data' variable already had 163 coord values defined for its date dimension. With the new values this dimension changes to length 29, so new coord values for the date dimension are required.`

It seems like the `set_data()` method might be missing from the MMM Class?

**Comment by williambdean** [comment]
Hello @louismagowan , have you come across the methods that are inherited from the `ModelBuilder`?
`predict_posterior` might be of interest to you. It will take advantage of the `set_data` functionality.
They are in the docs as well, for instance: https://www.pymc-marketing.io/en/stable/api/generated/pymc_marketing.mmm.delayed_saturated_mmm.DelayedSaturatedMMM.predict_posterior.html#pymc_marketing.mmm.delayed_saturated_mmm.DelayedSaturatedMMM.predict_posterior

**Reply by louismagowan** [reply]
Hi @wd60622 - I had tried using this function but got the following error:

`DTypePromotionError: The DType <class 'numpy.dtypes.DateTime64DType'> could not be promoted by <class 'numpy.dtypes.Float64DType'>. This means that no common DType exists for the given inputs. For example they cannot be stored in a single array unless the dtype is `object`. The full list of DTypes is: (<class 'numpy.dtypes.DateTime64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Float64DType'>)`


When I remove my date column from `X_test` I then receive the following error:

`ValueError: The 'channel_data' variable already had 163 coord values defined for its date dimension. With the new values this dimension changes to length 29, so new coord values for the date dimension are required.`


The code I'm using is the following:

```
# Create MMM
model = mmm.DelayedSaturatedMMM(date_column = "date", 
                                channel_columns = channel_cols,
                                adstock_max_lag = 4,
                                model_config = my_model_config,
                                sampler_config = sampler_config,
                                validate_data = True,
                                control_columns = control_cols,
                                yearly_seasonality = 10)

# Split into test and train
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    random_state=42, 
                                                    shuffle=False,
                                                    test_size = 0.15)

# Fit model on train data
model.fit(X=X_train, y=y_train,
          target_accept=0.95, 
          # Alter chains to specify how many cores to use, default uses 4 cores
          # Additional cores will not run faster - it will just do more chains
          chains=6,
          random_seed=42)

# Get test set predictions
test_preds = model.predict_posterior(X_test)
```

**Reply by williambdean** [reply]
Interesting. Do you have any null values in the `date` column? If so, are they `NaN` or `NaT`: https://stackoverflow.com/questions/71611309/how-to-solve-typeerror-the-dtypes-class-numpy-dtypeuint8-and-class-nump

Can you confirm that X_train has 163 rows and X_test has 29 rows

Also, you mention removing the `date` column but I don't see that in the example. Does the error message come from having `date` in `X_test` but removing it in `X_train`?

**Reply by louismagowan** [reply]
Have attached screenshot to hopefully answer your questions ☺️ 

`date` column is indeed present in both `X_train` and `X_test`. I didn't include the drop date but it's just 

```
# Get test set predictions
X_test = X_test.drop(columns = ["date"])
test_preds = model.predict_posterior(X_test)
```

<img width="484" alt="image" src="https://github.com/pymc-labs/pymc-marketing/assets/59659198/4ea8472d-761d-4ff8-8abd-e560ff03054f">

**Reply by williambdean** [reply]
Cool. Thanks for sharing that. 

I'm looking at the `_data_setter` method that calls the `pm.set_data`. It doesn't seem to set the fourier transformations or date. 
https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/mmm/delayed_saturated_mmm.py#L452-L512
Is that as intended @ricardoV94 ?

For a hacky solution, you should be able to work around this by building the model again with the X_test and y_test then sample using the first model's posterior.

I'm thinking this conceptually : 

```python
# Same as before 
model_settings = {...}
model_1 = DelayedSaturatedMMM(**model_settings)
model_1.fit(X_train, y_train)

# Hacky way
model_2 = DelayedSaturatedMMM(**model_settings)
# Processes the data and setup up the correct coords
model_2.build_model(X_test, y_test)

# Get the predictions using model_1.idata
with model_2.model: 
    new_idata = pm.sample_posterior_predictive(model_1.idata)
```

**Reply by ricardoV94** [reply]
May be a better question for @juanitorduz

**Reply by juanitorduz** [reply]
I think we currently do not support this but hopefully, we'll take it soon. As noticed above we need to take better care of the data containers and coordinates. PRs are always welcome :)

**Reply by kmostdiek-k** [reply]
FWIW, with the hacky/conceptual suggestion above, I get a `ValueError: date_col spend_week has repeated values`. I double checked and there shouldn't be any duplicates in the `spend_week` but maybe I'm missing something. 
![Screenshot 2023-10-20 at 11 20 21 AM](https://github.com/pymc-labs/pymc-marketing/assets/92057622/3f723be3-18fd-4cc7-8857-8723e7a18f6e)

**Reply by williambdean** [reply]
That error suggests that there are duplicates and is a check outside of PyMC. Are you able to check for duplicates to confirm, @kmostdiek-k?

`test_x[date_column].value_counts()` shows which are duplicates

**Reply by kmostdiek-k** [reply]
Yep, no duplicates.
![Screenshot 2023-10-20 at 11 56 07 AM](https://github.com/pymc-labs/pymc-marketing/assets/92057622/4f4879ff-1d77-46e7-af27-763152ac3d60)

**Reply by williambdean** [reply]
Try to initialize the class without `validate_data` then to get past the validate check. If not, I will take a deeper look

Also created issue #400 to address this

**Reply by kmostdiek-k** [reply]
I tried that `validate_data = False` and seem to have gotten around the repeat values error but now I'm getting an `AttributeError: 'Series' object has no attribute 'reshape'`.

**Reply by williambdean** [reply]
Could be from many things but sounds like a numpy.array methods being called on pandas objects. Trying passing in the target as numpy.array `y_test.to_numpy()`

**Reply by kmostdiek-k** [reply]
Sorry for the delayed response! I'm then getting a NaN error (see screenshot) but I've confirmed that I do not have any NaNs in either the X's or the Y's.
![Screenshot 2023-10-30 at 12 58 57 PM](https://github.com/pymc-labs/pymc-marketing/assets/92057622/8f4f38ca-1acb-4e39-8204-dc1b3c9e08ab)

**Reply by williambdean** [reply]
Interesting. Can you check for nan values in `mmm.preprocessed_data["X"]` rather than X and y
This seems related to the #402 

On top of that, there is an issue to correctly handle new data at the moment. I linked it above

**Reply by kmostdiek-k** [reply]
Yeah, looks like there are no NaN there either. Both of the following, result in `False`
```
mmm.preprocessed_data["X"].isnull().any().any()
np.isnan(np.min(mmm.preprocessed_data["y"]))
```
And yeah, I'm fine with waiting until either issue 400 or 396 are resolved. I just figured I'd help with the discussion here.

**Reply by williambdean** [reply]
There is another PR up to support the new data use case #444 
Hopefully will be merge sometime soon

**Reply by williambdean** [reply]
For anyone in the thread, a reworked PR (#482) is in the final stages of review

**Comment by juanitorduz** [comment, marked as answer]
@wd60622 did a fantastic job! https://github.com/pymc-labs/pymc-marketing/pull/482 . We will provide and example soon. Stay tuned!

**Reply by williambdean** [reply]
Example notebook up!
https://www.pymc-marketing.io/en/stable/notebooks/mmm/mmm_lift_test.html
